Replace non-public TabletClosedException and ScanTimedOutException

SethSmucker opened the pull request:
## Summary
- Replace `instanceof TabletClosedException` with class name check
- Replace `catch (ScanTimedOutException)` with class name check inside generic catch block

## Files Changed
- `QueryIterator.java` - TabletClosedException
- `DocIdQueryIterator.java` - TabletClosedException  
- `DocumentIterator.java` - TabletClosedException
- `Scan.java` - ScanTimedOutException

## Approach
Since these are non-public Accumulo classes that won't exist in Accumulo 4, we use class name comparison:
```java
private static boolean isTabletClosedException(Throwable t) {
    return t != null && t.getClass().getName().equals(
        "org.apache.accumulo.tserver.tablet.TabletClosedException");
}
```

Fixes #3317
Part of #2443

SethSmucker committed:
Replace non-public exception class checks with class name comparison

Replace instanceof checks for non-public Accumulo exceptions with
class name string comparison for Accumulo 4 compatibility:

- TabletClosedException in QueryIterator, DocIdQueryIterator, DocumentIterator
- ScanTimedOutException in Scan

Fixes #3317
Part of #2443

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/replace-nonpublic-exceptions

foster33 commented on warehouse/query-core/src/main/java/datawave/query/tables/async/Scan.java:
```diff
@@ -289,10 +288,13 @@ public Scan call() throws Exception {
+            if (isScanTimedOutException(e)) {
+                // this is okay. This means that we are being timesliced.
+                myScan.addRange(currentRange);
+                return this;
```
Should we be returning this here? Just curious because it was not in the original catch block you removed

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/iterator/QueryIterator.java:
```diff
@@ -576,22 +575,29 @@ else if (documentRange != null && (!this.isContainsIndexOnlyTerms() && this.getT
      *             for read/write issues
      */
+    /**
+     * Check if the throwable is a TabletClosedException by class name to avoid importing non-public Accumulo API.
```
I think it would be better for this function (along with the one for ScanTimedOutException to be extracted to a utility class so that we can avoid repeating code and better document why we're using string class names to check the class type.

SethSmucker committed:
Merge integration into task/replace-nonpublic-exceptions

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/replace-nonpublic-exceptions

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/replace-nonpublic-exceptions

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/replace-nonpublic-exceptions

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/replace-nonpublic-exceptions

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/replace-nonpublic-exceptions

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/replace-nonpublic-exceptions

SethSmucker committed:
Restructure ScanTimedOutException catch to match original control flow

Remove early return in favor of else branch, preserving the original
fall-through to return after finally block.

foster33 committed:
Merge branch 'integration' into task/replace-nonpublic-exceptions

ddanielr reviewed (changes requested):
Sorry but I would not recommend merging these changes.

While this current approach would work to eliminate the import, it's extremely fragile and will cause issues later.

#### main concern
This code is trying to ensure that if specific accumulo exceptions have been thrown they are raised to accumulo for proper handling instead of being handled by datawave code.

When/If these exception classes change in the future then these checks will always return false and accumulo exceptions will be swallowed during runtime.

#### Possible Alternative
A better approach might be to consolidate these accumulo exception type checks to a single class with a `checkExceptionType()` method and then just adding import control statements to limit the use of these exceptions to that class. 

This doesn't eliminate the import, but for this specific case I believe that's a better approach as it results in the failure condition being discovered at compile time instead of runtime.

ddanielr commented on warehouse/query-core/src/main/java/datawave/query/tables/async/Scan.java:
```diff
@@ -308,6 +310,13 @@ public Scan call() throws Exception {
+     * Check if the throwable is a ScanTimedOutException by class name to avoid importing non-public Accumulo API.
+     */
+    private boolean isScanTimedOutException(Throwable t) {
+        return t != null && t.getClass().getName().equals("org.apache.accumulo.core.clientImpl.ThriftScanner$ScanTimedOutException");
```
This class name doesn't look correct.

SethSmucker committed:
Add AccumuloExceptionChecker to centralize non-public exception checks

Replace fragile string-based class name comparisons with compile-safe
instanceof checks behind a centralized utility class, per ddanielr's
review feedback on PR #3318.

Non-public imports (TabletClosedException, ScanTimedOutException) are
now restricted to AccumuloExceptionChecker. If Accumulo renames or
removes these exceptions, the build fails immediately instead of
silently swallowing exceptions at runtime.

Part of #2443

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/replace-nonpublic-exceptions

SethSmucker committed:
Fix misplaced javadoc on handleException methods

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/replace-nonpublic-exceptions

SethSmucker committed:
Merge branch 'integration' into task/replace-nonpublic-exceptions

SethSmucker committed:
Add /etc/hadoop/conf symlink in quickstart Docker image

Datawave's server-side query iterators load Hadoop configuration from
/etc/hadoop/conf when they run inside the Accumulo TabletServer. That
path does not exist in the quickstart container — Hadoop is installed
under /opt/datawave/contrib/datawave-quickstart/hadoop and HADOOP_CONF_DIR
points there, not at /etc/hadoop/conf. Without the path, every
event-table scan throws FileNotFoundException server-side and the
query service silently returns 0 events.

Symlink /etc/hadoop/conf to the installed Hadoop config directory so the
expected path resolves. Verified locally: docker/scripts/testAll.sh goes
from 5/13 to 13/13 passing.

SethSmucker committed:
Add quickstart dependency for services that connect to ZooKeeper or Hadoop

Nine services in docker-compose.yml use ZOOKEEPER_HOST or HADOOP_HOST
but did not declare a depends_on for the quickstart container, which
hosts ZooKeeper, HDFS, and Accumulo. Only metrics correctly declared
the dependency.

When quickstart's startup happens to take longer than the 60s ZooKeeper
connect timeout used by Spring services like audit's AccumuloAuditor
bean, those services exit 1 during init with "Failed to connect to
zookeeper (quickstart:2181) within 2x zookeeper timeout period 30000",
and docker compose aborts with "dependency failed to start: container
docker-audit-1 exited (1)". The race is environment-dependent: on
slower runners it is frequently lost, producing intermittent CI failures.

Declare depends_on quickstart (condition: service_healthy) for
annotation, accumulo, audit, dictionary, modification, query,
mapreduce-query, executor-pool1, and executor-pool2 so the startup is
correctly serialized behind the container that provides their backing
services.

SethSmucker committed:
Merge branch 'integration' into task/compose-tests-stability

SethSmucker committed:
Bump spring-boot-starter-datawave-query consumers to 1.0.11-SNAPSHOT

PR #3560 removed deprecated eventPerDayThreshold and shardsPerDayThreshold
placeholders from the starter source and from docker/config/application-query.yml,
but the released spring-boot-starter-datawave-query-1.0.10.jar still embeds
the old QueryLogicFactory.xml that references those placeholders. With no
fresh 1.0.11 release cut yet, every consumer of the 1.0.10 jar (query,
executor, mapreduce-query, cached-results) fails Spring init at startup
with:

  Caused by: java.lang.IllegalArgumentException: Could not resolve
  placeholder 'datawave.query.logic.logics.BaseEventQuery.eventPerDayThreshold'

This blocks the compose-tests workflow on every recent PR opened against
integration (#3559, #3562, #3563, #3565 all show the same FAILURE).

Bump <version.datawave.starter-query> from 1.0.10 to 1.0.11-SNAPSHOT in the
six consuming poms so the multi-module reactor builds the starter from
current source (which has #3560's removal) and the query/executor docker
images embed the fresh jar.

This is a stop-gap. The proper fix is running microservice-cascade-release.yml
starting from microservices/starters/query to publish a real 1.0.11; once
that lands, these SNAPSHOT pins should be replaced with the released version.

SethSmucker committed:
Merge origin/integration into task/replace-nonpublic-exceptions

SethSmucker committed:
Merge task/compose-tests-stability into task/replace-nonpublic-exceptions for compose-tests CI fixes

SethSmucker committed:
Merge remote-tracking branch 'upstream/integration' into upd-3318

# Conflicts:
#	microservices/services/mapreduce-query/jobs/core/pom.xml
#	microservices/services/mapreduce-query/service/pom.xml
#	microservices/services/query-executor/api/pom.xml
#	microservices/services/query-executor/service/pom.xml
#	microservices/services/query/service/pom.xml
#	microservices/starters/cached-results/pom.xml
